core/drivers/fsl_sec/hw_key_blob.c

```diff
@@ -216,22 +222,69 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
 
 TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
 {
-	TEE_Result res;
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t temp_buffer[KEY_BLOB_SIZE];
+
+	COMPILE_TIME_ASSERT(KEY_BLOB_SIZE > HW_UNIQUE_KEY_LENGTH);
 
 	res = init_and_roll_forward_master_key();
 	if (res != TEE_SUCCESS) {
 		EMSG("Failed to init normal priblob otpmk identity from CAAM");
-		return res;
+		goto out;
 	}
 
-	res = get_hw_unq_key_blob_hw(&hwkey->data[0], HW_UNIQUE_KEY_LENGTH);
+	// Use the first 16 bytes for the HUK, the last 16 for the die ID.
+	res = get_hw_unq_key_blob_hw(temp_buffer, KEY_BLOB_SIZE);
 
-	if (res != TEE_SUCCESS) {
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	memcpy(&hwkey->data[0], temp_buffer, HW_UNIQUE_KEY_LENGTH);
+
+out:
+	if (res != TEE_SUCCESS)
 		EMSG("Hardware Unique Key Failed");
-	} else {
+	else
 		DMSG("Hardware Unique Key Retrieved:");
-		DHEXDUMP(&hwkey->data[0], HW_UNIQUE_KEY_LENGTH);
+
+	return res;
+}
+
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t temp_buffer[KEY_BLOB_SIZE];
+	uint8_t *remaining_buffer = &(temp_buffer[HW_UNIQUE_KEY_LENGTH]);
+	uint32_t remaining_size = KEY_BLOB_SIZE - HW_UNIQUE_KEY_LENGTH;
+
+	// Make sure there is some data left available after taking the HUK.
+	COMPILE_TIME_ASSERT(KEY_BLOB_SIZE > HW_UNIQUE_KEY_LENGTH);
+
+	if (len > TEE_SHA256_HASH_SIZE) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
 	}
 
+	res = init_and_roll_forward_master_key();
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to init normal priblob otpmk identity from CAAM");
+		goto out;
+	}
+
+	// Use the first 16 bytes for the HUK, the last 16 for the die ID.
+	res = get_hw_unq_key_blob_hw(temp_buffer, KEY_BLOB_SIZE);
+
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = tee_hash_createdigest(TEE_ALG_SHA256, remaining_buffer, remaining_size,
+				buffer, len);
+
+out:
+	if (res != TEE_SUCCESS)
+		EMSG("Die ID Failed");
+	else
+		DMSG("Die ID Retrieved:");
```

Comment:
nit: any reason Failed and Retrieved are capitalized?  Also the colon after retrieved suggests that something follows it, but there's not extra print statement.

Reply:
We used to dump the (super secret) value of the buffer. I removed that but missed the colon. I'll update it when I add the testing fallback which returns a zero buffer.